pkts-sip/src/main/java/io/pkts/packet/sip/impl/ImmutableSipMessage.java

```diff
@@ -319,33 +303,27 @@ public void verify() {
 
     }
 
-    private SipHeader findHeader(final Buffer name) {
-        for (final SipHeader header : headers) {
-            if (name.equals(header.getName())) {
-                return header;
-            }
-        }
-        return null;
+    private SipHeader findHeader(final String name) {
```

Comment:
why not keep the `Buffer` version? Which then would apply to have the map as `Map<Buffer, List<SipHeader>>`...

Reply:
Usually you search headers with strings like, "Via", "X-XPTO",.... so on.
If I use Buffer I would need to wrap every time I search an header.
I believe that's more cpu and memory required.

Reply:
well, perhaps from your particular use case but if you are close to the network, as an actual stack implementation would be, it deals with byte[] and as such, the Buffer wrapper.

A suggestion in your use case would be to define those as constans, which already eixsts on the various headers as well (and that you now did `toString()` on). So, just define the headers you look for in your app and it should be fine. I.e.

```
static final Buffer X_XPTO = Buffers.wrap(...);
```
etc.

Reply:
Trade off between parsing and searching. I can live with that :)
I've made some benchs, what you gain during parse you lose (considerably) when using getHeader.
I believe String based map has the best overall search speed (almos same performance with parse)

Benchmark                                       Mode  Cnt         Score        Error  Units
SIPHeadersBench.getBufferHeaderMapWithBuffers  thrpt   60  19962935,340 ± 417294,419  ops/s
SIPHeadersBench.getStringHeaderMapWithBuffers  thrpt   60   9739651,646 ± 298384,826  ops/s
SIPHeadersBench.parseMapWithBuffers            thrpt   60    164164,897 ±   2883,146  ops/s.

Benchmark                                      Mode  Cnt         Score        Error  Units
SIPHeadersBench.getBufferHeaderMapWithString   thrpt   60   9762874,386 ± 192812,414  ops/s
SIPHeadersBench.getStringHeaderMapWithString   thrpt   60  27756092,332 ± 291778,284  ops/s
SIPHeadersBench.parseMapWithString             thrpt   60    154003,602 ±   1461,894  ops/s

Reply:
nice, love your data driven approach! Hard to argue against it :-)

One more comment/concern/curiosity that I had but I think it got lost in the resolved comments - what about GC performance. With the lists in the maps, we are arguably creating more garbage, which may, or may not, affect long running programs.

Any chance you can run a test that runs for some time (hours at least, preferably days but I can do that later too). After that, I'll shut up and merge :-)

Reply:
:P
I can run that later (If you wouldn't mind), I'm on a tight schedule and I would prefer not to maintain an additional repo when we release a product version (which is happening very soon)
Though I don't believe i would make a difference, even depending on GC strategy it would be negligible